Adding example for using transactions phoenix spark application.

chrajeshbabu opened the pull request.

stoty commented on phoenix-spark-transactions/pom.xml:
```diff
@@ -0,0 +1,122 @@
+      <groupId>org.apache.phoenix</groupId>
+      <artifactId>phoenix5-spark</artifactId>
+      <!-- Phoenix spark connector version given by COD -->
+      <version>6.0.0.7.2.14.0-133</version>
```
Actually, none of the  dependencies that have downstream versions are needed, they should just be dropped.

chrajeshbabu commented on phoenix-spark-transactions/README.md:
```diff
@@ -0,0 +1,189 @@
+         
+      ],
+      "files":[
+        "hbase-shaded-mapreduce-2.4.6.7.2.14.0-133.jar",
```
Removed unnecessary jars.

chrajeshbabu commented on phoenix-spark-transactions/spark-app.json:
```diff
@@ -0,0 +1,30 @@
+
+      ],
+      "files":[
+        "hbase-shaded-mapreduce-<hbase.version>.jar",
```
This file is only required for CDE mentioned it's document only. Currently hbase-site.xml added as dependency in CDE job configuration is not picking so currently adding the configs to project resources and building.

chrajeshbabu commented on phoenix-spark-transactions/src/main/scala/com/cloudera/cod/examples/spark/SparkApp.scala:
```diff
@@ -0,0 +1,126 @@
+    var extraOptions = "phoenix.transactions.enabled=true,phoenix.upsert.batch.size=100";
+    df.write
+      .format("phoenix")
+      .options(Map("table" -> tableName, PhoenixDataSource.ZOOKEEPER_URL -> zkUrl,
```
Would be better to specify explicitly to show case that transactions enabled in the applications otherwise the configs specifying files is ok.

chrajeshbabu commented on phoenix-spark-transactions/src/main/scala/com/cloudera/cod/examples/spark/SparkApp.scala:
```diff
@@ -0,0 +1,126 @@
+    try {
+      conn.commit()
+    } catch {
+      case e: Exception => println("This is expected because of trying to commit conflicting data set.")
```
Used specific exception.

chrajeshbabu commented on phoenix-spark-transactions/pom.xml:
```diff
@@ -0,0 +1,122 @@
+    <scala.compat.version>2.11</scala.compat.version>
+    <spec2.version>4.2.0</spec2.version>
+    <!-- update it to the DEX supported version -->
+    <spark.version>2.4.0</spark.version>
```
Used the downstream versions.

chrajeshbabu commented on phoenix-spark-transactions/README.md:
````diff
@@ -0,0 +1,189 @@
+```
+cde resource upload --name spark-app-resource --local-path ./target/connector-libs/hbase-shaded-mapreduce-2.4.6.7.2.14.0-133.jar --resource-path hbase-shaded-mapreduce-2.4.6.7.2.14.0-133.jar
+cde resource upload --name spark-app-resource --local-path ./target/connector-libs/opentelemetry-api-0.12.0.jar --resource-path opentelemetry-api-0.12.0.jar
+cde resource upload --name spark-app-resource --local-path ./target/connector-libs/phoenix5-spark-6.0.0.7.2.14.0-133.jar --resource-path phoenix5-spark-6.0.0.7.2.14.0-133.jar
````
Removed it.

chrajeshbabu commented on phoenix-spark-transactions/README.md:
```diff
@@ -0,0 +1,189 @@
+cde resource upload --name spark-app-resource --local-path ./target/connector-libs/opentelemetry-api-0.12.0.jar --resource-path opentelemetry-api-0.12.0.jar
+cde resource upload --name spark-app-resource --local-path ./target/connector-libs/phoenix5-spark-6.0.0.7.2.14.0-133.jar --resource-path phoenix5-spark-6.0.0.7.2.14.0-133.jar
+cde resource upload --name spark-app-resource --local-path ./target/connector-libs/phoenix5-spark-shaded-6.0.0.7.2.14.0-133.jar --resource-path phoenix5-spark-shaded-6.0.0.7.2.14.0-133.jar
+cde resource upload --name spark-app-resource --local-path ./target/connector-libs/phoenix-client-hbase-2.4-5.1.1.7.2.14.0-133.jar --resource-path phoenix-client-hbase-2.4-5.1.1.7.2.14.0-133.jar
```
removed the unnecessary jars.

chrajeshbabu commented on phoenix-spark-transactions/README.md:
```diff
@@ -0,0 +1,189 @@
+<project>
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.phoenix</groupId>
```
Removed the unnecessary jars.

chrajeshbabu commented on phoenix-spark-transactions/src/main/scala/com/cloudera/cod/examples/spark/SparkApp.scala:
```diff
@@ -0,0 +1,126 @@
+      .format("phoenix")
+      .options(Map("table" -> tableName, PhoenixDataSource.ZOOKEEPER_URL -> zkUrl,
+        PhoenixDataSource.PHOENIX_CONFIGS -> extraOptions))
+      .mode(SaveMode.Overwrite)
```
Made it SaveMode.Append anyway.

chrajeshbabu commented:
Thank you @stoty  for reviews.Handled the review comments. Please take a look at it.

stoty commented on phoenix-spark-transactions/README.CDP.md:
````diff
@@ -0,0 +1,54 @@
+
+
+```
+spark-submit --master local[2] --jars /opt/cloudera/parcels/CDH/lib/phoenix_connectors/phoenix5-spark-shaded-6.0.0.7.2.14.0-117.jar --class com.cloudera.cod.examples.spark.SparkApp ./target/phoenix-spark-transactions-0.1.0.jar jdbc:phoenix:[***zookeeper_quorum***]:[***zookeeper_port***]:[***zookeeper_hbase_path***]
````
If you followed the docs, then you don't need to add the connector JAR.

chrajeshbabu replied:
we don't need connector jar removed it.

stoty commented on README.md:
```diff
@@ -34,3 +34,15 @@ various company's stock prices.
+A Scala application which creates a Phoenix transactional table, writes some rows
+in batch wise transactions, partition wise transactions and demonstrates
+how conflicts between streaming applications and spark applications are handled to ensure consitency.
+
```
Add something like:
This is the same application as the CDP Private Cloude Base example, but illustrates the differences between packaging and running the Spark application in CDP Private Cloude Base and in Cloudera Data Engineering products.

chrajeshbabu replied:
done.

stoty commented on README.md:
```diff
@@ -34,3 +34,15 @@ various company's stock prices.
+
+A Scala application which creates a Phoenix transactional table, writes some rows
+in batch wise transactions, partition wise transactions and demonstrates
+how conflicts between streaming applications and spark applications are handled to ensure consitency.
```
Add something like:
This is the same application as the Cloudera Data Engineering example, but illustrates the differences between packaging and running the Spark application in CDP Private Cloude Base and in Cloudera Data Engineering products.

stoty commented on README.md:
```diff
@@ -34,3 +34,15 @@ various company's stock prices.
+in batch wise transactions, partition wise transactions and demonstrates
+how conflicts between streaming applications and spark applications are handled to ensure consitency.
+
+* [Apache Phoenix Spark Connector, Spark, CDP DC, Maven: Transactions](phoenix-spark-transactions/README.CDP.md)
```
We call CDP DC "CDP Private Cloude Base" now.

chrajeshbabu replied:
changed it.

stoty commented on phoenix-spark-transactions/README.CDP.md:
```diff
@@ -0,0 +1,54 @@
+# Prerequisites
+
+1. Ensure Phoenix and Spark Installed In CDP-DC
```
Do not call it CDP-DC.

chrajeshbabu replied:
changed it.

stoty commented on phoenix-spark-transactions/README.md:
````diff
@@ -0,0 +1,152 @@
+```
+<properties>
+...
+    <phoenix.connector.version>6.0.0.7.2.14.0-133</phoenix.connector.version>
````
You don't need the phoenix connector version

chrajeshbabu replied:
Removed it.

stoty commented on phoenix-spark-transactions/src/main/scala/com/cloudera/cod/examples/spark/SparkApp.scala:
```diff
@@ -0,0 +1,126 @@
+import org.apache.spark.sql.{Row, SaveMode, SparkSession}
+import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
+import org.apache.omid.tso.client.AbortException
+import org.apache.phoenix.spark.datasource.v2.PhoenixDataSource
```
Don't depend on the datasource.

chrajeshbabu replied:
Removed this dependency.

stoty commented on phoenix-spark-transactions/src/main/scala/com/cloudera/cod/examples/spark/SparkApp.scala:
```diff
@@ -0,0 +1,126 @@
+    var extraOptions = "phoenix.transactions.enabled=true,phoenix.upsert.batch.size=100";
+    df.write
+      .format("phoenix")
+      .options(Map("table" -> tableName, PhoenixDataSource.ZOOKEEPER_URL -> zkUrl,
```
Use "zkUrl" and "phoenixconfigs".

PhoenixDataSource is an implementation detail.
It is also not present in the upcoming Spark3 drivers.

Spark + connector works hard to hide the implementation details, let's not add an avoidable dependencey for the sake of importing two String contants.

chrajeshbabu replied:
Removed the dependency.

stoty commented on phoenix-spark-transactions/pom.xml:
```diff
@@ -8,12 +8,11 @@
-    <scala.binary.version>2.12</scala.binary.version>
+    <spark.version>2.4.0</spark.version>
+    <scala.binary.version>2.11</scala.binary.version>
     <phoenix.connector.version>6.0.0.7.2.14.0-133</phoenix.connector.version>
```
The maven dependency is still present.

stoty reviewed (changes requested):
We don't need any of the Hbase or Phoenix dependencies in the Maven project.
Those are runtime dependencies, and are added by Spark.

chrajeshbabu commented:
> We don't need any of the Hbase or Phoenix dependencies in the Maven project.

Just added the dependencies to download so that those can be uploaded to CDE while creating job. Just added comment.

stoty reviewed (approved):
+1 LGTM
